README.md
# Essentia API Server

Simple python3 API server to provide a HTTP API to retrieve similar tracks to
any provided seed tracks.


## Analysing Tracks

Before this script can function you first need to anayze your tracks using
[Essentia Analyzer](https://github.com/CDrummond/essentia-analyzer)


## Similarity API 

The API server can be installed as a Systemd service, or started manually:

```
./essentia-api.py -c config.json -l DEBUG
```

Only 1 API is currently supported:

```
http://HOST:11000/api/similar?track=/path/of/track&track=/path/of/another/track&count=10&filtergenre=1&min=30&max=600&norepart=15&norepalb=25&previous=/path/to/previous&filterxmas=1&shuffle=1
```
...this will get 10 similar tracks to those supplied.

If `filtergenre=1` is supplied then only tracks whose genre matches a
pre-configured set of genres (mapped from seed tracks) will be used. e.g. if
`["Heavy Metal", "Metal", "Power Metal"]` is defined in the config, and a seed
tack's genre has `Metal` then only tracks with one of these 3 genres will be
considered.

If `filterxmas=1` is supplied, then tracks with 'Christmas' or 'Xmas' in their
genres will be excluded - unless it is December.

`min` and `max` can be used to set the minimum, and maximum, duration (in
seconds) of tracks to be considered.

`norepart` specifies the number of tracks where an artist should not be
repeated. This is not a hard-limit, as if there are too few candidates then
repeats can happen.

`norepalb` specifies the number of tracks where an album should not be
repeated. This does not aply to 'Various Artist' albums. This is not also not a
hard-limit, as if there are too few candidates then repeats can happen.

`previous` may be used to list tracks to ignore (e.g. tracks that are already in
the queue). This parameter, like `track`, may be repeated multiple times.

`shuffle` indicates that the chosen tracks should be shuffled.

For each seed track, the API will attempt to locate the desired `count` similar
tracks.  If `shuffle=1` is supplied then double `count` number of tracks will
be found, these wll be shuffled, and then then top `count` tracks returned.

Metadata for tracks is stored in an SQLite database, this has an `ignore` column
which if set to `1` will cause the API to not use this track if it is returned
as a similar track by essentia. In this way you can exclude specific tracks from
being added to mixes - but if they are already in the queue, then they can sill
be used as seed tracks.

This API is intended to be used by [LMS Music Similarity Plugin](https://github.com/CDrummond/lms-musicsimilarity)

Genres are configured via the `genres` section of `config.json`, using the
following syntax:

```
{
 "genres:[
  [ "Rock", "Hard Rock", "Metal" ],
  [ "Pop", "Dance", "R&B"]
 ]
}
```

If a seed track has `Hard Rock` as its genre, then only tracks with `Rock`,
`Hard Rock`, or `Metal` will be allowed. If a seed track has a genre that is not
listed here then any track returned by Musly, that does not cotaiain any genre
lsited here, will be considered acceptable. Therefore, if seed is `Pop` then
a `Hard Rock` track would not be considered.

### HTTP Post

Alternatively, the API may be accessed via a HTTP POST call. To do this, the
params of the call are passed as a JSON object. eg.

```
{
 "track":["/path/trackA.mp3", "/path/trackB.mp3"],
 "filtergenre":1,
 "count":10
}
```

## Configuration

The sever reads its configuration from a JSON file (default name is `config.json`).
This has the following format:

```
{
 "lms":"/home/storage/Music/",
 "db":"/home/craig/Development/Essentia/lms-essentia/essentia.db",
 "genres":[
  ["Alternative Rock", "Classic Rock", "Folk/Rock", "Hard Rock", "Indie Rock", "Punk Rock", "Rock"],
  ["Dance", "Disco", "Hip-Hop", "Pop", "Pop/Folk", "Pop/Rock", "R&B", "Reggae", "Soul", "Trance"],
  ["Gothic Metal", "Heavy Metal", "Power Metal", "Progressive Metal", "Progressive Rock", "Symphonic Metal", "Symphonic Power Metal"]
 ],
 "ignoregenre":["Artist"],
 "normalize":{
  "artist":["feet", "ft", "featuring"],
  "album":["deluxe edition", "remastered"],
  "title"["demo", "radio edit"]
 },
 "port":11000,
 "host":"0.0.0.0"
}
```

* `lms` should be the path where LMS access your music files. The API
server will remove this path from API calls, so that it can look up tracks in
its database by their relative path.
* `db` should contain the path to the Essentia DB containing the analysis results.
* `genres` This is as described above.
* `ignoregenre` List of artists where genre filtering (excluding christmas)
should be ignored. To apply to all artists, use '*' - e.g. `"ignoregenre":"*"`
* `normalize.artist` List of strings to split artist names, e.g. "A ft. B"
becomes "A" (periods are automatically removed)
* `normalize.album` List of strings to remove from album names.
* `normalize.title` List of strings to remove from titles.
* `port` This is the port number the API is accessible on.
* `host` IP addres on which the API will listen on. Use `0.0.0.0` to listen on
all interfaces on your network.



ChangeLog
0.0.2
-----
1. Allow configuring no repeat artist/album/track
2. Fix parsing of 1/0 parameters.
3. Expand /api/dump to allow usage in MusicSimilarity LMS plugin.

0.0.1
-----
1. Initial release.


mkrel.py
#!/usr/bin/env python3

#
# Essentia API Server
#
# Copyright (c) 2020-2021 Craig Drummond <craig.p.drummond@gmail.com>
# GPLv3 license.
#

import hashlib
import os
import re
import requests
import shutil
import subprocess
import sys


APP_NAME = "essentia-api"

def info(s):
    print("INFO: %s" %s)


def error(s):
    print("ERROR: %s" % s)
    exit(-1)


def usage():
    print("Usage: %s <major>.<minor>.<patch>" % sys.argv[0])
    exit(-1)


def checkVersion(version):
    try:
        parts=version.split('.')
        major=int(parts[0])
        minor=int(parts[1])
        patch=int(parts[2])
    except:
        error("Invalid version number")


def releaseUrl(version):
    return "https://github.com/CDrummond/%s/releases/download/%s/%s-%s.zip" % (APP_NAME, version, APP_NAME, version)


def checkVersionExists(version):
    url = releaseUrl(version)
    info("Checking %s" % url)
    request = requests.head(url)
    if request.status_code == 200 or request.status_code == 302:
        error("Version already exists")


def updateVersion(version):
    path = os.path.join('lib', 'version.py')
    os.remove(path)
    with open(path, "w") as f:
        f.write("ESSENTIA_API_VERSION='%s'\n" % version)


def resetVersion():
    subprocess.call(['git', 'checkout', os.path.join('lib', 'version.py')], shell=False)


def createZip(version):
    info("Creating ZIP")
    os.chdir('..')
    cmd=["zip", "-r", "%s/%s-%s.zip" % (APP_NAME, APP_NAME, version)]

    for f in ["ChangeLog", "config.json", "essentia-api.py", "essentia-api.service", "LICENSE", "README.md"]:
        cmd.append("%s/%s" % (APP_NAME, f))
    for e in os.listdir("%s/lib" % APP_NAME):
        if e.endswith(".py"):
            cmd.append("%s/lib/%s" % (APP_NAME, e))
    subprocess.call(cmd, shell=False)
    os.chdir(APP_NAME)


version=sys.argv[1]
if version!="test":
    checkVersion(version)
    checkVersionExists(version)
    updateVersion(version)

createZip(version)

if version!="test":
    resetVersion();


lib/config.py
#
# Essentia API Service for LMS
#
# Copyright (c) 2020-2021 Craig Drummond <craig.p.drummond@gmail.com>
# GPLv3 license.
#

import json
import logging
import os
from . import tracks_db

_LOGGER = logging.getLogger(__name__)

def read_config(path):
    config={}

    if not os.path.exists(path):
        _LOGGER.error('%s does not exist' % path)
        exit(-1)
    try:
        with open(path, 'r') as configFile:
            config = json.load(configFile)
    except ValueError:
        _LOGGER.error('Failed to parse config file')
        exit(-1)
    except IOError:
        _LOGGER.error('Failed to read config file')
        exit(-1)

    for key in ['lms', 'db']:
        if not key in config:
            _LOGGER.error("'%s' not in config file" % key)
            exit(-1)
        if (key=='db' and not os.path.exists(config[key])):
            _LOGGER.error("'%s' does not exist" % config['db'])
            exit(-1)

    if not 'port' in config:
        config['port']=11002

    if not 'host' in config:
        config['host']='0.0.0.0'

    if 'genres' in config:
        config['all_genres']=[]
        for genres in config['genres']:
            for g in genres:
                if not g in config['all_genres']:
                    config['all_genres'].append(g)

    if 'ignoregenre' in config:
        if isinstance(config['ignoregenre'], list):
            ignore=[]
            for item in config['ignoregenre']:
                ignore.append(tracks_db.normalize_artist(item))
            config['ignoregenre']=ignore
        else:
            config['ignoregenre']=[config['ignoregenre']]

    if 'normalize' in config:
        tracks_db.set_normalize_options(config['normalize'])

    return config


lib/filters.py
#
# Analyse files with Musly, and provide an API to retrieve similar tracks
#
# Copyright (c) 2020-2021 Craig Drummond <craig.p.drummond@gmail.com>
# GPLv3 license.
#

VARIOUS_ARTISTS = ['various', 'various artists'] # Artist names are normalised, and coverted to lower case
CHRISTMAS_GENRES = ['Christmas', 'Xmas']

def same_artist_or_album(seeds, track, check_album_only=False, max_check=0):
    check = 0
    for seed in seeds:
        if seed['artist']==track['artist'] and not check_album_only:
            return True
        if seed['album']==track['album'] and 'albumartist' in seed and 'albumartist' in track and seed['albumartist']==track['albumartist'] and track['albumartist'] not in VARIOUS_ARTISTS:
            return True
        check+=1
        if max_check>0 and check>=max_check:
            return False
    return False


def genre_matches(config, seed_genres, track):
    if 'igenres' not in track or len(track['igenres'])<1:
        return True # Track has no genre? Then can't filter out...

    # Ignore genre for an artist?
    if 'ignoregenre' in config and track['artist'] in config['ignoregenre']:
        return True

    if len(seed_genres)<1:
        # No filtering for seed track genres
        if 'all_genres' in config:
            for tg in track['igenres']:
                if tg in config['all_genres']:
                    # Track's genre is in config list, but not in seeds, so filter out track
                    return False
        # No seed genres, and track's genre not in filters, so accept track
        return True

    for sg in seed_genres:
        if sg in track['igenres']:
            return True

    return False


def is_christmas(track):
    if 'genres' in track and len(track['genres'])>=0:
        for genre in track['genres']:
            if genre in CHRISTMAS_GENRES:
                return True

    return False


def match_artist(artists, track):
    for artist in artists:
        if artist==track['artist'] or ('albumartist' in track and artist==track['albumartist']):
            return True
    return False


def match_album(albums, track):
    if (not 'album' in track) or ( ('albumartist' not in track) and ('artist' not in track)):
        return False

    album = '%s - %s' % (track['albumartist'] if 'albumartist' in track else track['artist'], track['album'])
    return album in albums


def match_title(titles, track):
    if not 'title' in track:
        return False

    return track['title'] in titles


def check_duration(min_duration, max_duration, meta):
    if 'duration' not in meta or meta['duration'] is None or meta['duration']<=0:
        return True # No duration to check!

    if min_duration>0 and meta['duration']<min_duration:
        return False

    if max_duration>0 and meta['duration']>max_duration:
        return False

    return True



essentia-api.py
#!/usr/bin/env python3

#
# Analyse files with essentia, and provide an API to retrieve similar tracks
#
# Copyright (c) 2020-2021 Craig Drummond <craig.p.drummond@gmail.com>
# GPLv3 license.
#

import argparse
import logging
import os
from lib import app, config, version

_LOGGER = logging.getLogger(__name__)
        
if __name__=='__main__':
    parser = argparse.ArgumentParser(description='Essentia API Server (v%s)' % version.ESSENTIA_API_VERSION)
    parser.add_argument('-c', '--config', type=str, help='Config file (default: config.json)', default='config.json')
    parser.add_argument('-l', '--log-level', action='store', choices=['CRITICAL', 'ERROR', 'WARNING', 'INFO', 'DEBUG'], default='INFO', help='Set log level (default: %(default)s)')
    args = parser.parse_args()
    logging.basicConfig(format='%(asctime)s %(levelname)-8s %(message)s', level=args.log_level, datefmt='%Y-%m-%d %H:%M:%S')
    cfg = config.read_config(args.config)
    app.start_app(args, cfg)



lib/app.py
#
# Essentia API Service for LMS
#
# Copyright (c) 2020-2021 Craig Drummond <craig.p.drummond@gmail.com>
# GPLv3 license.
#

import argparse
from datetime import datetime
import json
import logging
import os
import random
import sqlite3
import time
import urllib
from flask import Flask, abort, request
from . import cue, filters, tracks_db


_LOGGER = logging.getLogger(__name__)

DEFAULT_TRACKS_TO_RETURN              = 5    # Number of tracks to return, if none specified
MIN_TRACKS_TO_RETURN                  = 5    # Min value for 'count' parameter
MAX_TRACKS_TO_RETURN                  = 50   # Max value for 'count' parameter
DEFAULT_NUM_PREV_TRACKS_FILTER_ARTIST = 15   # Try to ensure artist is not in previous N tracks
DEFAULT_NUM_PREV_TRACKS_FILTER_ALBUM  = 25   # Try to ensure album is not in previous N tracks
SHUFFLE_FACTOR                        = 3    # How many (shuffle_factor*count) tracks to shuffle?
VERBOSE_DEBUG                         = True


class EssentiaApp(Flask):
    def init(self, args, app_config):
        _LOGGER.debug('Start server')
        self.app_config = app_config

        flask_logging = logging.getLogger('werkzeug')
        flask_logging.setLevel(args.log_level)
        flask_logging.disabled = 'DEBUG'!=args.log_level
        # Load tracks DB into memory
        tracks_db.TracksDb(app_config)

        random.seed()

    def get_config(self):
        return self.app_config
    
essentia_app = EssentiaApp(__name__)


def get_value(params, key, defVal, isPost):
    if isPost:
        return params[key] if key in params else defVal
    return params[key][0] if key in params else defVal


def decode(url, root):
    u = urllib.parse.unquote(url)
    if u.startswith('file://'):
        u=u[7:]
    elif u.startswith('tmp://'):
        u=u[6:]
    if u.startswith(root):
        u=u[len(root):]
    return cue.convert_from_cue_path(u)


def log_track(reason, track):
    if VERBOSE_DEBUG:
        _LOGGER.debug('%s %s' % (reason, str(track)))


@essentia_app.route('/api/dump', methods=['GET', 'POST'])
def dump_api():
    isPost = False
    if request.method=='GET':
        params = request.args.to_dict(flat=False)
    else:
        isPost = True
        params = request.get_json()
        _LOGGER.debug('Request: %s' % json.dumps(params))

    if not params:
        abort(400)

    if not 'track' in params:
        abort(400)

    if len(params['track'])!=1:
        abort(400)

    cfg = essentia_app.get_config()
    db = tracks_db.TracksDb(cfg)

    # Strip LMS root path from track path
    root = cfg['lms']

    track = decode(params['track'][0], root)
    entry = db.get(track, True)
    if entry is None:
        abort(404)

    fmt = get_value(params, 'format', '', isPost)
    match_artist = int(get_value(params, 'filterartist', '0', isPost))==1

    tracks = db.get_similar_tracks(entry,  match_all_genres=1==int(get_value(params, 'matchallgenres', '0', isPost)), count=-1)
    count = int(get_value(params, 'count', 1000, isPost))

    if not fmt.startswith('text'):
        return json.dumps(tracks[:count])

    resp=[]
    _LOGGER.debug('Num tracks %d' % len(tracks))
    if fmt=='text-url':
        tracks.insert(0, entry)
        for track in tracks:
            if match_artist and entry['artist']!=track['artist']:
                continue
            path = '%s%s' % (root, track['file'])
            resp.append(cue.convert_to_cue_url(path))
            if len(resp)>=count:
                break
    else:
        header = "file\tsimilarity\tgenres"

        if fmt=='textall':
            for attr in tracks_db.ESSENTIA_ATTRIBS:
                header+="\t%s" % attr
        resp.append(header)
        for track in tracks:
            if match_artist and entry['artist']!=track['artist']:
                continue
            if 'genres' in track:
                line="%s\t%f\t%s" % (track['file'], track['similarity'], track['genres'])
            else:
                line="%s\t%f\tXXXXX" % (track['file'], track['similarity'])
            if fmt=='textall':
                for attr in tracks_db.ESSENTIA_ATTRIBS:
                    line+="\t%f" % track[attr]
            resp.append(line)
            if len(resp)>=count:
                break

    return '\n'.join(resp)


@essentia_app.route('/api/similar', methods=['GET', 'POST'])
def similar_api():
    tstart = time.time_ns()
    isPost = False
    if request.method=='GET':
        params = request.args.to_dict(flat=False)
    else:
        isPost = True
        params = request.get_json()
        _LOGGER.debug('Request: %s' % json.dumps(params))

    if not params:
        abort(400)

    if not 'track' in params:
        abort(400)

    count = int(get_value(params, 'count', DEFAULT_TRACKS_TO_RETURN, isPost))
    if count < MIN_TRACKS_TO_RETURN:
        count = MIN_TRACKS_TO_RETURN
    elif count > MAX_TRACKS_TO_RETURN:
        count = MAX_TRACKS_TO_RETURN

    match_genre = int(get_value(params, 'filtergenre', '0', isPost))==1
    shuffle = int(get_value(params, 'shuffle', '1', isPost))==1
    min_duration = int(get_value(params, 'min', 0, isPost))
    max_duration = int(get_value(params, 'max', 0, isPost))
    no_repeat_artist = int(get_value(params, 'norepart', 0, isPost))
    no_repeat_album = int(get_value(params, 'norepalb', 0, isPost))
    exclude_christmas = int(get_value(params, 'filterxmas', '0', isPost))==1 and datetime.now().month!=12

    if no_repeat_artist<0 or no_repeat_artist>200:
        no_repeat_artist = DEFAULT_NUM_PREV_TRACKS_FILTER_ARTIST
    if no_repeat_album<0 or no_repeat_album>200:
        no_repeat_album = DEFAULT_NUM_PREV_TRACKS_FILTER_ALBUM

    cfg = essentia_app.get_config()
    db = tracks_db.TracksDb(cfg)

    # Strip LMS root path from track path
    root = cfg['lms']
    
    # Similar tracks
    similar_tracks=[]
    # Similar tracks ignored because of artist/album
    filtered_by_seeds_tracks=[]
    filtered_by_current_tracks=[]
    filtered_by_previous_tracks=[]
    current_titles=[]

    # Set of rows from seeds/previous, and already checked items
    skip_rows=set()

    if min_duration>0 or max_duration>0:
        _LOGGER.debug('Duration:%d .. %d' % (min_duration, max_duration))

    seed_track_db_entries=[]
    seed_genres=set()
    for trk in params['track']:
        track = decode(trk, root)
        _LOGGER.debug('S TRACK %s -> %s' % (trk, track))

        # Check that we know about this track
        entry = db.get(track, True)
        if entry is not None:
            seed_track_db_entries.append(entry)
            skip_rows.add(entry['rowid'])
            if 'igenres' in entry and 'genres' in cfg:
                for genre in entry['igenres']:
                    for group in cfg['genres']:
                        if genre in group:
                            for cg in group:
                                if not cg in seed_genres:
                                    seed_genres.add(cg)
            if 'title' in entry:
                current_titles.append(entry['title'])
        else:
            _LOGGER.debug('Could not locate %s in DB' % track)

    previous_track_db_entries = []
    if 'previous' in params:
        for trk in params['previous']:
            track = decode(trk, root)
            _LOGGER.debug('P TRACK %s -> %s' % (trk, track))

            entry = db.get(track, False)
            if entry is not None:
                previous_track_db_entries.append(entry)
                if entry['rowid'] not in skip_rows:
                    skip_rows.add(entry['rowid'])
                if 'title' in entry:
                    current_titles.append(entry['title'])
        _LOGGER.debug('Have %d previous tracks to ignore' % len(previous_track_db_entries))

    if match_genre:
        _LOGGER.debug('Seed genres: %s' % seed_genres)

    similarity_count = int(count * SHUFFLE_FACTOR) if shuffle else count

    _LOGGER.debug('Setup time:%d' % int((time.time_ns()-tstart)/1000000))

    matched_artists={}
    for seed in seed_track_db_entries:
        accepted_tracks = 0
        match_all_genres = ('ignoregenre' in cfg) and ('*'==cfg['ignoregenre'][0] or (seed['artist'] in cfg['ignoregenre']))

        # Query DB for similar tracks
        resp = db.get_similar_tracks(seed, match_all_genres, len(skip_rows))

        for track in resp:

            if track['rowid'] in skip_rows or (min_duration>0 and track['duration']<min_duration) or (max_duration>0 and track['duration']>max_duration):
                continue

            if match_genre and not match_all_genres and not filters.genre_matches(cfg, seed_genres, track):
                log_track('DISCARD(genre)', track)
            elif exclude_christmas and filters.is_christmas(track):
                log_track('DISCARD(xmas)', track)
            else:
                if filters.same_artist_or_album(seed_track_db_entries, track):
                    log_track('FILTERED(seeds)', track)
                    filtered_by_seeds_tracks.append(track)
                elif filters.same_artist_or_album(similar_tracks, track):
                    log_track('FILTERED(current)', track)
                    filtered_by_current_tracks.append(track)
                    if track['artist'] in matched_artists and track['similarity'] - matched_artists[track['artist']]['similarity'] <= 0.25:
                        matched_artists[track['artist']]['tracks'].append(track)
                elif no_repeat_artist>0 and filters.same_artist_or_album(previous_track_db_entries, track, False, no_repeat_artist):
                    log_track('FILTERED(previous(artist))', track)
                    filtered_by_previous_tracks.append(track)
                elif no_repeat_album>0 and filters.same_artist_or_album(previous_track_db_entries, track, True, no_repeat_album):
                    log_track('FILTERED(previous(album))', track)
                    filtered_by_previous_tracks.append(track)
                elif filters.match_title(current_titles, track):
                    log_track('FILTERED(title)', track)
                    filtered_by_previous_tracks.append(track)
                else:
                    log_track('USABLE', track)
                    similar_tracks.append(track)
                    # Keep list of all tracks of an artist, so that we can randomly select one => we don't always use the same one
                    matched_artists[track['artist']]={'similarity':track['similarity'], 'tracks':[track], 'pos':len(similar_tracks)-1}
                    if 'title' in track:
                        current_titles.append(track['title'])
                    skip_rows.add(track['rowid'])
                    accepted_tracks += 1
                    if accepted_tracks >= similarity_count:
                        break


    # For each matched_artists randonly select a track...
    for matched in matched_artists:
        if len(matched_artists[matched]['tracks'])>1:
            if VERBOSE_DEBUG:
                _LOGGER.debug('Choosing random track for %s (%d tracks)' % (matched, len(matched_artists[matched]['tracks'])))
            sim = similar_tracks[matched_artists[matched]['pos']]['similarity']
            similar_tracks[matched_artists[matched]['pos']] = random.choice(matched_artists[matched]['tracks'])
            similar_tracks[matched_artists[matched]['pos']]['similarity'] = sim

    # Too few tracks? Add some from the filtered lists
    min_count = 2
    if len(similar_tracks)<min_count and len(filtered_by_previous_tracks)>0:
        _LOGGER.debug('Add some tracks from filtered_by_previous_tracks, %d/%d' % (len(similar_tracks), len(filtered_by_previous_tracks)))
        filtered_by_previous_tracks = sorted(filtered_by_previous_tracks, key=lambda k: k['similarity'])
        similar_tracks = similar_tracks + filtered_by_previous_tracks[:min_count-len(similar_tracks)]
    if len(similar_tracks)<min_count and len(filtered_by_current_tracks)>0:
        _LOGGER.debug('Add some tracks from filtered_by_current_tracks, %d/%d' % (len(similar_tracks), len(filtered_by_current_tracks)))
        filtered_by_current_tracks = sorted(filtered_by_current_tracks, key=lambda k: k['similarity'])
        similar_tracks = similar_tracks + filtered_by_current_tracks[:min_count-len(similar_tracks)]
    if len(similar_tracks)<min_count and len(filtered_by_seeds_tracks)>0:
        _LOGGER.debug('Add some tracks from filtered_by_seeds_tracks, %d/%d' % (len(similar_tracks), len(filtered_by_seeds_tracks)))
        filtered_by_seeds_tracks = sorted(filtered_by_seeds_tracks, key=lambda k: k['similarity'])
        similar_tracks = similar_tracks + filtered_by_seeds_tracks[:min_count-len(similar_tracks)]

    # Sort by similarity
    similar_tracks = sorted(similar_tracks, key=lambda k: k['similarity'])
    
    # Take top 'similarity_count' tracks
    similar_tracks = similar_tracks[:similarity_count]

    if shuffle:
        random.shuffle(similar_tracks)
        similar_tracks = similar_tracks[:count]

    track_list = []
    for track in similar_tracks:
        path = '%s%s' % (root, track['file'])
        track_list.append(cue.convert_to_cue_url(path))
        if VERBOSE_DEBUG:
            _LOGGER.debug('Path:%s %f' % (path, track['similarity']))

    db.close()
    _LOGGER.debug('Total time:%d' % int((time.time_ns()-tstart)/1000000))
    if get_value(params, 'format', '', isPost)=='text':
        return '\n'.join(track_list)
    else:
        return json.dumps(track_list)


def start_app(args, config):
    essentia_app.init(args, config)
    _LOGGER.debug('Ready to process requests')
    essentia_app.run(host=config['host'], port=config['port'])
